### PR TITLE
CFY-2214 Add key file

### PR DESCRIPTION
**Pull request by szpotona**

### CODE REVIEW

File: cloudify_hostpool/config/yaml_pool.py [hunk 1]
```diff
@@ -76,7 +77,11 @@ def _create_host(_host):
                     "'host' or the 'ip_range' key")
 
     def _get_auth(self, host):
-        return host.get('auth') or self.default.get('auth')
+        auth = host.get('auth') or self.default.get('auth', {})
+        if auth.get('keyfile') and not os.access(auth['keyfile'], os.R_OK):
```
Comment (iliapolo) [inline cloudify_hostpool/config/yaml_pool.py]: the string 'keyfile' is repeated three times in this method.
its best just to do it once
keyfile = auth.get('keyfile) and use keyfile variable

File: cloudify_hostpool/config/yaml_pool.py [hunk 2]
```diff
@@ -76,7 +77,11 @@ def _create_host(_host):
...
+            raise exceptions.ConfigurationError(
+                'There is no key file: {0}'.format(auth['keyfile']))
```
Comment (iliapolo) [inline cloudify_hostpool/config/yaml_pool.py]: The message should reflect a permissions problem as well, corresponding to the conditions. 
maybe "keyfile {0} does not exist or does not have the proper permissions"

File: cloudify_hostpool/rest/backend.py [hunk 3]
```diff
@@ -125,3 +128,16 @@ def _is_alive(host):
         address, port = host['host'], host['port']
         results = scan.scan([(address, port)])
         return results[address, port]
+
+    @staticmethod
+    def _load_keyfile(host):
+        if not host['auth']:
```
Comment (iliapolo) [inline cloudify_hostpool/rest/backend.py]: this means the keyfile will be read on every request. maybe it better to do this at the configuration level, where the check for existence already takes place anyway

File: cloudify_hostpool/tests/config/test_yaml_pool.py [hunk 4]
```diff
@@ -184,3 +185,36 @@ def test_empty_config(self):
...
+        try:
+            pool = yaml_pool.YAMLPoolLoader(config)
+            self.assertRaisesRegexp(exceptions.ConfigurationError,
+                                    'There is no key file:', pool.load().next)
+        finally:
+            os.rmdir(tmp_dir)
+
+    def test_good_keyfile(self):
+        fd, good_file = tempfile.mkstemp()
+        os.close(fd)
```
Comment (iliapolo) [inline cloudify_hostpool/tests/config/test_yaml_pool.py]: why is this needed? the os.close i mean..mkstemp creates a brand new tempfile

File: cloudify_hostpool/tests/rest/test_backend.py [hunk 5]
```diff
@@ -43,12 +45,16 @@ class RestBackendTest(unittest.TestCase):
 
     def setUp(self):
         pool = {'hosts': self._generate_hosts(self.NUMBER_OF_HOSTS)}
-        self.backend = RestBackend(pool=pool, db_file_name=tempfile.mktemp())
+        fd, self.db_tmp = tempfile.mkstemp()
+        os.close(fd)
```
Comment (iliapolo) [inline cloudify_hostpool/tests/rest/test_backend.py]: same here

File: cloudify_hostpool/config/yaml_pool.py (hunk 1 above)
Comment (szpotona) [inline cloudify_hostpool/config/yaml_pool.py]: Good point.

File: cloudify_hostpool/config/yaml_pool.py (hunk 2 above)
Comment (szpotona) [inline cloudify_hostpool/config/yaml_pool.py]: +1

File: cloudify_hostpool/rest/backend.py (hunk 3 above)
Comment (msoczewka) [inline cloudify_hostpool/rest/backend.py]: You mean to cache key contents in memory upon configuration? Holding them in the database would be an antipattern.

Comment (iliapolo) [inline cloudify_hostpool/rest/backend.py]: you mean its fround upon to hold keys in the database?

File: cloudify_hostpool/tests/config/test_yaml_pool.py (hunk 4 above)
Comment (szpotona) [inline cloudify_hostpool/tests/config/test_yaml_pool.py]: According to tempfile documentation: https://docs.python.org/2/library/tempfile.html this method: tempfile.mkstemp() is safe and returns tuple(file_descriptor, file_path) file_descriptor which you have to close and file_path of this temporary file, thus I you delete it after usage.

File: cloudify_hostpool/tests/rest/test_backend.py (hunk 5 above)
Comment (szpotona) [inline cloudify_hostpool/tests/rest/test_backend.py]: According to tempfile documentation: https://docs.python.org/2/library/tempfile.html the method you've used: tempfile.mktemp() is obsolete and insecure. That's why I used tempfile.mkstemp()

File: cloudify_hostpool/tests/config/test_yaml_pool.py (hunk 4 above)
Comment (msoczewka) [inline cloudify_hostpool/tests/config/test_yaml_pool.py]: "_have to close_" = the function also returns an open file descriptor (because you usually want to use the temporary file you just have created).

File: cloudify_hostpool/rest/backend.py (hunk 3 above)
Comment (msoczewka) [inline cloudify_hostpool/rest/backend.py]: Exactly.

I would either cache the contents in memory or hold filesystem paths (which is not that bad solution).

Comment (iliapolo) [inline cloudify_hostpool/rest/backend.py]: ok, im fine with storing file system paths. but i think check for existence here is redundant, it was already validated during initial load

File: cloudify_hostpool/tests/config/test_yaml_pool.py (hunk 4 above)
Comment (iliapolo) [inline cloudify_hostpool/tests/config/test_yaml_pool.py]: ok

File: cloudify_hostpool/tests/rest/test_backend.py (hunk 5 above)
Comment (iliapolo) [inline cloudify_hostpool/tests/rest/test_backend.py]: ok

File: cloudify_hostpool/tests/rest/test_backend.py [hunk 6]
```diff
@@ -160,3 +166,33 @@ def _acquire():
...
+            self.backend = RestBackend(pool={'hosts': [host_with_key, ]},
+                                       db_file_name=new_db.name)
+            host = self.backend.acquire_host()
+            self.assertEqual(host['auth']['keyfile'], key_mock)
+            host = self.backend.get_host(host['host_id'])
+            self.assertEqual(host['auth']['keyfile'], key_mock)
+        finally:
+            if os.path.exists(keyfile):
+                os.unlink(keyfile)
+            self.backend = old_backend
```
Comment (iliapolo) [inline cloudify_hostpool/tests/rest/test_backend.py]: you dont have to do this. these tests initialize a rest backend on every test

Comment (iliapolo) [inline cloudify_hostpool/tests/rest/test_backend.py]: also, you dont have to initialize a new instance. if you want to ensure you acquire a specific host. just use backend.storage to manually acquire all other hosts.

File: cloudify_hostpool/rest/backend.py [hunk 7]
```diff
@@ -125,3 +127,13 @@ def _is_alive(host):
...
```
Comment (iliapolo) [inline cloudify_hostpool/rest/backend.py]: why is this needed? auth must provided and its validated in the config loading.

Comment (szpotona) [inline cloudify_hostpool/rest/backend.py]: That's true. It's redundant.
